include/sparrow/nullable.hpp

```diff
@@ -0,0 +1,589 @@
+            not impl::initializable_from_refs<T, nullable<TO, BO>>
+        )
+        explicit(not impl::both_convertible_from_cond_ref<T, TO, B, BO>)
+        constexpr nullable(nullable<TO, BO>&& rhs)
+            : m_value(std::move(rhs).get())
+            , m_flag(std::move(rhs).has_value())
+        {
+        }
+
+        constexpr nullable(value_type&& value, flag_type&& flag)
+            : m_value(std::move(value))
+            , m_flag(std::move(flag))
+        {
+        }
+
+        constexpr nullable(std::add_lvalue_reference_t<T> value, std::add_lvalue_reference_t<B> flag)
+            : m_value(value)
+            , m_flag(flag)
+        {
+        }
+
+        constexpr nullable(value_type&& value, std::add_lvalue_reference_t<B> flag)
+            : m_value(std::move(value))
+            , m_flag(flag)
+        {
+        }
+
+        constexpr nullable(std::add_lvalue_reference_t<T> value, flag_type&& flag)
+            : m_value(value)
+            , m_flag(std::move(flag))
+        {
+        }
+
+        constexpr self_type& operator=(nullval_t)
+        {
+            m_flag = false;
+            return *this;
+        }
+
+        template <class TO>
+        requires(
+            not std::same_as<self_type, TO> and
+            std::assignable_from<std::add_lvalue_reference_t<T>, TO>
+        )
+        constexpr self_type& operator=(TO&& rhs)
+        {
+            m_value = std::forward<TO>(rhs);
+            m_flag = true;
+            return *this;
+        }
+
+        constexpr self_type& operator=(const self_type& rhs)
+        {
+            m_value = rhs.get();
+            m_flag = rhs.has_value();
+            return *this;
+        }
+
+        template <class TO, boolean_like BO>
+        requires(
+            impl::both_assignable_from_cref<T, TO, B, BO> and
+            not impl::initializable_from_refs<T, nullable<TO, BO>> and
+            not impl::assignable_from_refs<T, nullable<TO, BO>>
+        )
+        constexpr self_type& operator=(const nullable<TO, BO>& rhs)
+        {
+            m_value = rhs.get();
+            m_flag = rhs.has_value();
+            return *this;
+        }
+
+        constexpr self_type& operator=(self_type&& rhs)
+        {
+            m_value = std::move(rhs).get();
+            m_flag = std::move(rhs).has_value();
+            return *this;
+        }
+
+        template <class TO, boolean_like BO>
+        requires(
+            impl::both_assignable_from_cond_ref<T, TO, B, BO> and
+            not impl::initializable_from_refs<T, nullable<TO, BO>> and
+            not impl::assignable_from_refs<T, nullable<TO, BO>>
+        )
+        constexpr self_type& operator=(nullable<TO, BO>&& rhs)
+        {
+            m_value = std::move(rhs).get();
+            m_flag = std::move(rhs).has_value();
+            return *this;
+        }
+
+        constexpr explicit operator bool() const noexcept;
+
+        constexpr flag_reference has_value() & noexcept;
```

Comment:
Having a way to change the flag only is harmless since there is no associated logic for destructing the value. Besides, these methods are required to correclty implement the different constructors.

However, I agree that the name is misleading. We could have a const has_value returning a bool, and rename these functions into "null_flag" for instance;

Reply:
That would make more sense indeed 👍🏽 

> Having a way to change the flag only is harmless since there is no associated logic for destructing the value. 

That's assuming we will not need any other kind of code in there in the future, I wouldn't try to guess myself haha

> Besides, these methods are required to correclty implement the different constructors.

Could you clarify this point? Arent the constructors already able to access the private members anyway?

Reply:
when moving / copying a nullable, you want to copy or move its underlying value and flag depending on whether the parameter is an lvalue or an rvalue, and depending on its nature: is it a nullable holding value or a nullable holding references. The different overloads of `values` and `null_flag` return the correct type in every situation.

Besides, constructor can indeed access the private members of another instance of the same type. But we also use these methods for conversion constructors and probably for efficient computation (when the number of null values in an array is low for instance).